man/man1/wsaver.1
.TH WSAVER 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wsaver \- screen saver for W
.SH SYNOPSIS
wsaver \fIseconds\fP
.SH DESCRIPTION
.I wsaver
is the new - external - screen saver. If you start it you must give a numeric
argument which is the number of seconds before it should get activated. Note
that this number must be greater than 10.
.PP
After the given time has expired without any user activity, \fIwsaver\fP
will activate one of it's 'modules' at random.  If you'll give a number
smaller than 10, \fIwsaver\fP will demo module corresponding to that
number in the root window.  Without a number \fIwsaver\fP shows the
available modules.
.PP
You can naturally have only one screen saver running at a time, and
since wsaver creates its window only when it gets activated it can also
fail due to memory problems. In this case the screen will simply
stay as it is, so you don't need to worry.
.SH LIMITATIONS
\'Ant-walk' module will use colors with a color server if there are
enough shared colors.  Other modules don't use colors.
.PP
There are no module specific options.
.SH CHANGED
15th Feb 1996 by TeSche
.br
26th Feb 1998 by Eero Tamminen, added five new modules.


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/wsaver/wsaver.c
/*
 * wsaver.c, a part of the W Window System
 *
 * Copyright (C) 1998 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a simple W screen saver
 *
 * CHANGES
 * 2/98, ++eero:
 * - The saver window is created here, drawn black and then exported.
 * - Runs in the root window if delay is less than 10 seconds.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/time.h>
#include <time.h>
#include <Wlib.h>
#include "wsaver.h"


struct saver_func {
  void (*func)(void);
  const char *name;
};

#define SAVERS	7

static struct saver_func saver[SAVERS] = {
  { save_ants,   "ants"   },
  { save_bounce, "bounce" },
  { save_cracks, "cracks" },
  { save_lines,  "lines"  },
  { save_pyro,   "pyro"   },
  { save_snails, "snails" },
  { save_stars,  "stars"  }
};


#define TIMEOUT 70	/* milliseconds between updates */

/* exported variables */
long timeout = TIMEOUT;
short swidth, sheight, scolors;
WWIN *win;


static void usage(void)
{
  int i;
  fprintf(stderr, "usage: wsaver <seconds>\n");
  fprintf(stderr, "       where seconds > %d\n\n", MAX(10,SAVERS));
  fprintf(stderr, "If you'll give a smaller number, a saver corresponding\n");
  fprintf(stderr, "to that number will be run in the root window.\n");
  for (i = 0; i < SAVERS; i++) {
    fprintf(stderr, "\t%d: %s\n", i, saver[i].name);
  }
}

int main(int argc, char *argv[])
{
  short seconds;
  WSERVER *wserver;
  WEVENT *ev;

  if (argc != 2) {
    usage();
    return -1;
  }

  seconds = atoi(argv[1]);

  if (!(wserver = w_init())) {
    fprintf(stderr, "%s: can't connect to wserver\n", *argv);
    return -1;
  }

  srandom(time(NULL));
  swidth = wserver->width;
  sheight = wserver->height;
  scolors = wserver->sharedcolors;

  if (seconds > MAX(10,SAVERS)) {
    if (w_setsaver(seconds) < 0) {
      fprintf(stderr, "%s: can't install screensaver\n", *argv);
      fprintf(stderr, "maybe you have already installed one?\n");
      w_exit();
      return -1;
    }
  } else {
    win = WROOT;
    w_setlinewidth(win, 1);		/* just in case... */
    w_setmode(win, M_DRAW);
    w_pbox(win, 0, 0, swidth, sheight);
    /* run wsaver in the root window now! */
    saver[seconds%SAVERS].func();
    w_exit();
    return 0;
  }

  while (42) {

    if ((ev = w_queryevent(NULL, NULL, NULL, -1))) {

      switch (ev->type) {

        case EVENT_GADGET:
	  if (ev->key == GADGET_EXIT) {
	    w_exit();
	    return 0;
	  }
	  break;

	case EVENT_SAVEON:
          win = w_create(swidth, sheight, W_TOP|W_NOBORDER|W_NOMOUSE);
	  if (!win) {
	    fprintf(stderr, "%s: unable to create a W_TOP window\n", *argv);
	    w_exit();
	    return -1;
          }
	  /* savers should run with most of the screen black... */
	  w_pbox(win, 0, 0, swidth, sheight);
	  w_open(win, 0, 0);

	  saver[random()%SAVERS].func();
	  w_delete(win);
	  break;
      }
    }
  }
}
